File: Sources/Infrastructure/CLI/CopilotUsageProbe.swift
import Foundation
import Domain
import Mockable
import os.log

private let logger = Logger(subsystem: "com.claudebar", category: "CopilotUsageProbe")

/// Protocol for providing GitHub credentials
@Mockable
public protocol GitHubCredentialProvider: Sendable {
    func getToken() -> String?
    func getUsername() -> String?
}

/// Default credential provider that uses KeychainService
public struct KeychainCredentialProvider: GitHubCredentialProvider {
    public init() {}

    public func getToken() -> String? {
        KeychainService.shared.getToken(for: KeychainService.Account.githubCopilotToken)
    }

    public func getUsername() -> String? {
        UserDefaults.standard.string(forKey: "githubUsername")
    }
}

/// Probe for fetching GitHub Copilot usage data via GitHub Billing API.
///
/// Uses the GitHub REST API to fetch premium request usage:
/// `GET /users/{username}/settings/billing/usage`
///
/// Requires a fine-grained PAT with "Plan: read" permission.
public struct CopilotUsageProbe: UsageProbe {
    private let networkClient: any NetworkClient
    private let credentialProvider: any GitHubCredentialProvider
    private let timeout: TimeInterval

    private static let apiBaseURL = "https://api.github.com"
    private static let apiVersion = "2022-11-28"

    public init(
        networkClient: any NetworkClient = URLSession.shared,
        credentialProvider: any GitHubCredentialProvider = KeychainCredentialProvider(),
        timeout: TimeInterval = 30
    ) {
        self.networkClient = networkClient
        self.credentialProvider = credentialProvider
        self.timeout = timeout
    }

    public func isAvailable() async -> Bool {
        guard let token = credentialProvider.getToken(),
              let username = credentialProvider.getUsername(),
              !token.isEmpty,
              !username.isEmpty else {
            logger.debug("Copilot: Not available - missing token or username")
            return false
        }
        return true
    }

    public func probe() async throws -> UsageSnapshot {
        guard let token = credentialProvider.getToken(), !token.isEmpty else {
            logger.error("Copilot: No GitHub token configured")
            throw ProbeError.authenticationRequired
        }

        guard let username = credentialProvider.getUsername(), !username.isEmpty else {
            logger.error("Copilot: No GitHub username configured")
            throw ProbeError.executionFailed("GitHub username not configured")
        }

        logger.debug("Copilot: Fetching billing usage for \(username)")

        // Fetch billing usage
        let usageData = try await fetchBillingUsage(username: username, token: token)

        // Parse and create snapshot
        return try parseUsageResponse(usageData, username: username)
    }

    // MARK: - API Calls

    private func fetchBillingUsage(username: String, token: String) async throws -> BillingUsageResponse {
        // Get current month's usage
        let now = Date()
        let calendar = Calendar.current
        let year = calendar.component(.year, from: now)
        let month = calendar.component(.month, from: now)

        let urlString = "\(Self.apiBaseURL)/users/\(username)/settings/billing/usage?year=\(year)&month=\(month)"

        guard let url = URL(string: urlString) else {
            throw ProbeError.executionFailed("Invalid URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue(Self.apiVersion, forHTTPHeaderField: "X-GitHub-Api-Version")
        request.timeoutInterval = timeout

        let (data, response) = try await networkClient.request(request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw ProbeError.executionFailed("Invalid response")
        }

        logger.debug("Copilot API response status: \(httpResponse.statusCode)")

        switch httpResponse.statusCode {
        case 200:
            break
        case 401:
            logger.error("Copilot: Authentication failed (401)")
            throw ProbeError.authenticationRequired
        case 403:
            logger.error("Copilot: Forbidden - check token permissions (403)")
            throw ProbeError.executionFailed("Forbidden - ensure PAT has 'Plan: read' permission")
        case 404:
            logger.error("Copilot: User not found or no billing access (404)")
            throw ProbeError.executionFailed("User not found or no billing access")
        default:
            logger.error("Copilot: HTTP error \(httpResponse.statusCode)")
            throw ProbeError.executionFailed("HTTP error: \(httpResponse.statusCode)")
        }

        do {
            return try JSONDecoder().decode(BillingUsageResponse.self, from: data)
        } catch {
            logger.error("Copilot: Failed to parse response - \(error.localizedDescription)")
            // Log raw response for debugging
            if let rawString = String(data: data, encoding: .utf8) {
                logger.debug("Raw response: \(rawString.prefix(500))")
            }
            throw ProbeError.parseFailed("Failed to parse billing response: \(error.localizedDescription)")
        }
    }

    // MARK: - Response Parsing

    private func parseUsageResponse(_ response: BillingUsageResponse, username: String) throws -> UsageSnapshot {
        // Filter for Copilot usage items
        let copilotItems = response.usageItems.filter { item in
            item.product?.lowercased().contains("copilot") == true
        }

        logger.debug("Copilot: Found \(copilotItems.count) Copilot usage items")

        // Calculate totals
        var totalQuantity: Double = 0
        var totalNetAmount: Double = 0
        var totalDiscountAmount: Double = 0

        for item in copilotItems {
            totalQuantity += item.quantity ?? 0
            totalNetAmount += item.netAmount ?? 0
            totalDiscountAmount += item.discountAmount ?? 0
        }

        // Calculate included units from discount
        // Assumes a price per unit to derive included quota
        let pricePerUnit: Double = 0.04 // Premium request price (approximate)
        let includedUnits = totalDiscountAmount > 0 ? totalDiscountAmount / pricePerUnit : 2500 // Default free tier

        // Calculate percentage remaining
        let totalAllowed = includedUnits
        let used = totalQuantity
        let remaining = max(0, totalAllowed - used)
        let percentRemaining = totalAllowed > 0 ? (remaining / totalAllowed) * 100 : 100

        logger.debug("Copilot: Used \(Int(used))/\(Int(totalAllowed)) requests, \(Int(percentRemaining))% remaining")

        // Create quota
        let quota = UsageQuota(
            percentRemaining: percentRemaining,
            quotaType: .session,
            providerId: "copilot",
            resetText: "Resets monthly"
        )

        return UsageSnapshot(
            providerId: "copilot",
            quotas: [quota],
            capturedAt: Date(),
            accountEmail: username
        )
    }
}

// MARK: - API Response Models

private struct BillingUsageResponse: Decodable {
    let usageItems: [UsageItem]

    enum CodingKeys: String, CodingKey {
        case usageItems = "usage_items"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.usageItems = (try? container.decode([UsageItem].self, forKey: .usageItems)) ?? []
    }
}

private struct UsageItem: Decodable {
    let date: String?
    let product: String?
    let sku: String?
    let quantity: Double?
    let unitType: String?
    let pricePerUnit: Double?
    let grossAmount: Double?
    let discountAmount: Double?
    let netAmount: Double?

    enum CodingKeys: String, CodingKey {
        case date, product, sku, quantity
        case unitType = "unit_type"
        case pricePerUnit = "price_per_unit"
        case grossAmount = "gross_amount"
        case discountAmount = "discount_amount"
        case netAmount = "net_amount"
    }
}

File: shared/infrastructure/src/commonMain/kotlin/com/tddworks/claudebar/infrastructure/probes/copilot/CopilotUsageProbe.kt
package com.tddworks.claudebar.infrastructure.probes.copilot

import com.tddworks.claudebar.domain.model.ProbeError
import com.tddworks.claudebar.domain.model.QuotaType
import com.tddworks.claudebar.domain.model.UsageQuota
import com.tddworks.claudebar.domain.model.UsageSnapshot
import com.tddworks.claudebar.domain.provider.CredentialKey
import com.tddworks.claudebar.domain.provider.CredentialRepository
import com.tddworks.claudebar.domain.provider.UsageProbe
import com.tddworks.claudebar.infrastructure.network.NetworkClient
import com.tddworks.claudebar.infrastructure.network.NetworkResponse
import kotlinx.datetime.Clock
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Probe for fetching GitHub Copilot usage data via GitHub Billing API.
 *
 * Uses the GitHub REST API to fetch premium request usage:
 * `GET /users/{username}/settings/billing/premium_request/usage`
 *
 * Requires a fine-grained PAT with "Plan: read" permission.
 */
class CopilotUsageProbe(
    private val networkClient: NetworkClient,
    private val credentialRepository: CredentialRepository,
    private val json: Json = Json { ignoreUnknownKeys = true }
) : UsageProbe {

    companion object {
        private const val API_BASE_URL = "https://api.github.com"
        private const val API_VERSION = "2022-11-28"
        private const val MONTHLY_LIMIT = 2000.0
    }

    override suspend fun isAvailable(): Boolean {
        val token = credentialRepository.get(CredentialKey.GITHUB_TOKEN)
        val username = credentialRepository.get(CredentialKey.GITHUB_USERNAME)
        return !token.isNullOrEmpty() && !username.isNullOrEmpty()
    }

    override suspend fun probe(): UsageSnapshot {
        val token = credentialRepository.get(CredentialKey.GITHUB_TOKEN)
        if (token.isNullOrEmpty()) {
            throw ProbeError.AuthenticationRequired
        }

        val username = credentialRepository.get(CredentialKey.GITHUB_USERNAME)
        if (username.isNullOrEmpty()) {
            throw ProbeError.ExecutionFailed("GitHub username not configured")
        }

        val usageData = fetchBillingUsage(username, token)
        return parseUsageResponse(usageData, username)
    }

    private suspend fun fetchBillingUsage(username: String, token: String): PremiumRequestUsageResponse {
        val url = "$API_BASE_URL/users/$username/settings/billing/premium_request/usage"

        val headers = mapOf(
            "Authorization" to "Bearer $token",
            "Accept" to "application/vnd.github+json",
            "X-GitHub-Api-Version" to API_VERSION
        )

        val response: NetworkResponse = networkClient.get(url, headers)

        when (response.statusCode) {
            200 -> {} // Continue
            401 -> throw ProbeError.AuthenticationRequired
            403 -> throw ProbeError.ExecutionFailed("Forbidden - ensure PAT has 'Plan: read' permission")
            404 -> throw ProbeError.ExecutionFailed("User not found or no billing access")
            else -> throw ProbeError.ExecutionFailed("HTTP error: ${response.statusCode}")
        }

        return try {
            json.decodeFromString<PremiumRequestUsageResponse>(response.body)
        } catch (e: Exception) {
            throw ProbeError.ParseFailed("Failed to parse billing response: ${e.message}")
        }
    }

    private fun parseUsageResponse(response: PremiumRequestUsageResponse, username: String): UsageSnapshot {
        val copilotItems = response.usageItems.filter { item ->
            item.product?.lowercase()?.contains("copilot") == true
        }

        val totalGrossQuantity = copilotItems.sumOf { it.grossQuantity ?: 0.0 }
        val used = totalGrossQuantity
        val remaining = maxOf(0.0, MONTHLY_LIMIT - used)
        val percentRemaining = (remaining / MONTHLY_LIMIT) * 100

        val quota = UsageQuota(
            percentRemaining = percentRemaining,
            quotaType = QuotaType.Session,
            providerId = "copilot",
            resetText = "${used.toInt()}/${MONTHLY_LIMIT.toInt()} requests"
        )

        return UsageSnapshot(
            providerId = "copilot",
            quotas = listOf(quota),
            capturedAt = Clock.System.now(),
            accountEmail = username
        )
    }
}

// MARK: - API Response Models

@Serializable
internal data class PremiumRequestUsageResponse(
    @SerialName("time_period") val timePeriod: TimePeriod,
    val user: String,
    @SerialName("usage_items") val usageItems: List<PremiumUsageItem>
) {
    @Serializable
    data class TimePeriod(
        val year: Int,
        val month: Int
    )
}

@Serializable
internal data class PremiumUsageItem(
    val product: String? = null,
    val sku: String? = null,
    val model: String? = null,
    @SerialName("unit_type") val unitType: String? = null,
    @SerialName("price_per_unit") val pricePerUnit: Double? = null,
    @SerialName("gross_quantity") val grossQuantity: Double? = null,
    @SerialName("gross_amount") val grossAmount: Double? = null,
    @SerialName("discount_quantity") val discountQuantity: Double? = null,
    @SerialName("discount_amount") val discountAmount: Double? = null,
    @SerialName("net_quantity") val netQuantity: Double? = null,
    @SerialName("net_amount") val netAmount: Double? = null
)
